### PR TITLE
Wip/cragw/dell k2 fix write fw 1.9.27

### PR DESCRIPTION
**Pull request by CragW**
Type of pull request:

- [ ] New plugin (Please include [new plugin checklist](https://github.com/fwupd/fwupd/wiki/New-plugin-checklist))
- [x] Code fix
- [ ] Feature
- [ ] Documentation

### CODE REVIEW

File: plugins/dell-k2/fu-dell-k2-ec-hid.h [hunk 1]
```diff
@@ -13,6 +13,9 @@
 #define DELL_K2_EC_HID_CMD_FWUPDATE    0xAB
 #define DELL_K2_EC_HID_EXT_FWUPDATE    0x80
 #define DELL_K2_EC_HID_SUBCMD_FWUPDATE 0x00
+#define DELL_K2_EC_DEV_EC_CHUNK_SZ     160000
+#define DELL_K2_EC_DEV_ANY_CHUNK_SZ    180000
+#define DELL_K2_EC_DEV_NO_CHUNK_SZ     9999999
```
Comment (hughsie) [inline plugins/dell-k2/fu-dell-k2-ec-hid.h:18]: not sure about this; don't you want `G_MAXSIZE`?

File: plugins/dell-k2/fu-dell-k2-rmm.c [hunk 2]
```diff
@@ -25,29 +25,29 @@ fu_dell_k2_rmm_convert_version(FuDevice *device, guint64 version_raw)
...
+{
+	FuDevice *parent = NULL;
+
+	/* use version given by parent */
+	parent = fu_device_get_parent(device);
+	if (parent != NULL) {
+		guint32 rmm_version;
+
+		rmm_version = fu_dell_k2_ec_get_rmm_version(parent);
+		fu_device_set_version_raw(device, GUINT32_FROM_BE(rmm_version));
```
Comment (hughsie) [inline plugins/dell-k2/fu-dell-k2-rmm.c:39]: I don't think fu_dell_k2_ec_get_rmm_version() should be returning a BE-encoded version -- i.e. it should be doing GUINT32_FROM_BE and returning, not returning and expecting the caller to do GUINT32_FROM_BE

File: plugins/dell-k2/fu-dell-k2-ec-hid.h (hunk 1 above)
Comment (CragW) [reply plugins/dell-k2/fu-dell-k2-ec-hid.h:18]: Oh I should use it instead the ugly num, will fix.

File: plugins/dell-k2/fu-dell-k2-rmm.c [hunk 3]
```diff
@@ -127,7 +83,7 @@ fu_dell_k2_rmm_init(FuDellK2Rmm *self)
 	fu_device_add_icon(FU_DEVICE(self), "dock-usb");
 	fu_device_add_flag(FU_DEVICE(self), FWUPD_DEVICE_FLAG_UPDATABLE);
 	fu_device_add_flag(FU_DEVICE(self), FWUPD_DEVICE_FLAG_SIGNED_PAYLOAD);
-	fu_device_add_private_flag(FU_DEVICE(self), FU_DEVICE_PRIVATE_FLAG_USE_PROXY_FOR_OPEN);
+	fu_device_add_flag(FU_DEVICE(self), FWUPD_DEVICE_FLAG_INSTALL_SKIP_VERSION_CHECK);
```
Comment (hughsie) [inline plugins/dell-k2/fu-dell-k2-rmm.c:86]: this really feels like a slippery slope -- we're not actually going to be doing "did it work" checks if we use this flag.

File: plugins/dell-k2/fu-dell-k2-ec.c [hunk 4]
```diff
@@ -850,6 +926,7 @@ fu_dell_k2_ec_init(FuDellK2Ec *self)
 	fu_device_add_flag(FU_DEVICE(self), FWUPD_DEVICE_FLAG_USABLE_DURING_UPDATE);
 	fu_device_add_flag(FU_DEVICE(self), FWUPD_DEVICE_FLAG_DUAL_IMAGE);
 	fu_device_add_flag(FU_DEVICE(self), FWUPD_DEVICE_FLAG_SELF_RECOVERY);
+	fu_device_add_flag(FU_DEVICE(self), FWUPD_DEVICE_FLAG_INSTALL_SKIP_VERSION_CHECK);
```
Comment (hughsie) [inline plugins/dell-k2/fu-dell-k2-ec.c:929]: so the intention is not to do any end-to-end testing?

File: plugins/dell-k2/fu-dell-k2-ec.c [hunk 5]
```diff
@@ -672,75 +663,160 @@ fu_dell_k2_ec_commit_package(FuDevice *device, GBytes *blob_fw, GError **error)
...
+						      DELL_K2_EC_HID_DATA_PAGE_SZ);
+
+		/* iterate the pages */
+		for (guint j = 0; j < fu_chunk_array_length(pages); j++) {
+			guint8 packet[DELL_K2_EC_HID_DATA_PAGE_SZ] = {0x00};
+			g_autoptr(FuChunk) page = NULL;
+
+			page = fu_chunk_array_index(pages, j, error);
+			if (page == NULL)
+				return FALSE;
```
Comment (hughsie) [inline plugins/dell-k2/fu-dell-k2-ec.c:740]: if you do `fu_chunk_get_bytes()` you could use `fu_bytes_pad()` here I think.

File: plugins/dell-k2/fu-dell-k2-ec.c [hunk 6]
```diff
@@ -672,75 +663,160 @@ fu_dell_k2_ec_commit_package(FuDevice *device, GBytes *blob_fw, GError **error)
...
 		chk = fu_chunk_array_index(chunks, i, error);
 		if (chk == NULL)
 			return FALSE;
 
-		if (!fu_dell_k2_ec_hid_write(device, fu_chunk_get_bytes(chk), error))
-			return FALSE;
-	}
-	fu_progress_step_done(progress);
+		/* prepend header and command to the chunk data */
+		buf = fu_dell_k2_ec_hid_fwup_pkg_new(fu_chunk_get_bytes(chk),
```
Comment (hughsie) [inline plugins/dell-k2/fu-dell-k2-ec.c:722]: this leaks GBytes -- and I'm struggling to see why you need to pass in the GBytes, and the fz size too.

Comment (CragW) [reply plugins/dell-k2/fu-dell-k2-ec.c:722]: > this leaks GBytes

Sending the chunk directly would fix the leak issue?

> why you need to pass in the GBytes, and the fz size too.

The header needs to understand the size of a chunk, and the total size of a firmware binary.

File: plugins/dell-k2/fu-dell-k2-ec.c (hunk 5 above)
Comment (CragW) [reply plugins/dell-k2/fu-dell-k2-ec.c:740]: using `fu_bytes_pad()` with size `DELL_K2_EC_HID_DATA_PAGE_SZ` it is complaing a stack overflow issue.

```
*** stack smashing detected ***: terminated
```

File: plugins/dell-k2/fu-dell-k2-rmm.c (hunk 2 above)
Comment (CragW) [reply plugins/dell-k2/fu-dell-k2-rmm.c]: fixed.

File: plugins/dell-k2/fu-dell-k2-ec.c (hunk 4 above)
Comment (CragW) [reply plugins/dell-k2/fu-dell-k2-ec.c:927]: Yeah, for now the dock reboot command for online activation yet ready. I plan to add it later when it becomes available.

File: plugins/dell-k2/fu-dell-k2-plugin.c [hunk 7]
```diff
@@ -69,6 +69,31 @@ fu_dell_k2_plugin_device_add(FuPlugin *plugin, FuDevice *device, GError **error)
...
+	if (pid == DELL_K2_USB_RMM_PID) {
+		g_autoptr(FuDellK2Rmm) rmm_device = NULL;
+		g_autoptr(FuDeviceLocker) locker = NULL;
+
+		rmm_device = fu_dell_k2_rmm_new(FU_USB_DEVICE(device));
+		if (rmm_device == NULL) {
+			g_set_error(error,
+				    FWUPD_ERROR,
+				    FWUPD_ERROR_INTERNAL,
+				    "failed to create rtshub device, pid: 0x%04x",
```
Comment (superm1) [inline plugins/dell-k2/fu-dell-k2-plugin.c:82]: This error is wrong no?

Comment (CragW) [reply plugins/dell-k2/fu-dell-k2-plugin.c:82]: thanks for the catch, will fix.

**Review by superm1 — APPROVED** [review]

**Pull request merged**